### ISSUE TITLE
Odd UbSan warning - redc_p384 not of expected type

### ISSUE THREAD

**Issue by randombit** [randombit/botan#1370]
Clang 5.0 on x86-64 Linux.

I don't understand this warning. Commenting out the P-384 test in `unit_ecc.cpp`, it starts warning about the next run (P-256). Tried converting from `std::function` to a template parameter, no change.

```
nist_redc:
/bin/../lib64/gcc/x86_64-pc-linux-gnu/7.2.1/../../../../include/c++/7.2.1/bits/std_function.h:316:2: runtime error: call to function Botan::redc_p384(Botan::BigInt&, std::vector<unsigned long, Botan::secure_allocator<unsigned long> >&) through pointer to incorrect function
 type 'void (*)(Botan::BigInt &, std::vector<unsigned long, Botan::secure_allocator<unsigned long> > &)'
(/home/jack/work/botan/libbotan-2.so.4+0x15d0fb0): note: Botan::redc_p384(Botan::BigInt&, std::vector<unsigned long, Botan::secure_allocator<unsigned long> >&) defined here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /bin/../lib64/gcc/x86_64-pc-linux-gnu/7.2.1/../../../../include/c++/7.2.1/bits/std_function.h:316:2 in 
```

Maybe some bad interaction between Clang and `libstdc++`? Unfortunately `libc++` doesn't seem to work on my machine.

**Comment by randombit** [comment]
No error with GCC 7.2.1 UBsan

**Comment by mgierlings** [comment]
Should be fixed via #1373.
